enhancements/storage/csi-migration.md

```diff
@@ -0,0 +1,220 @@
+  2. Once that's done, disable the feature gate in all control-plane components.
+
+In order to keep the Attach Detach Controller and the Kubelet in sync regarding using the CSI driver or the in-tree plugin, upstream has a mechanism to keep the Attach Detach Controller informed about the status of the migration on nodes. Roughly speaking, Kubelet tracks this status via an annotation with the information for each migrated in-tree plugin on the node.
+
+As a result, the Attach Detach Controller knows if the in-tree plugin has been migrated on the node. If the feature flags are enabled in Kube Controller Manager **and** on the node, the Attach Detach Controller uses the CSI driver to attach volumes. Otherwise, it will revert to the in-tree plugin.
+
+#### Upstream and OCP patches
+
+In OCP, we can easily set those feature gates by using the [FeatureGate] (https://docs.openshift.com/container-platform/4.7/nodes/clusters/nodes-cluster-enabling-features.html) Custom Resource. OCP operators read this resource and restart their operands with the appropriate features enabled.
+However, this approach alone is not acceptable for CSI migration because the feature flags might be switched across components in _any_ arbitrary order. Our solution is intended to make the Attach Detach Controller resilient to this issue.
+
+That being said, **we propose to carry a patch in OCP's Attach Controller to force-enable the CSI Migration feature gates there**.
+
+This carry-patch would only affect the Attach Detach Controller. Other parts of Kube Controller Manager, such as the PersistentVolume Controller, would still obey the flags passed to the Kube Controller Manager.
+In order to do this the least invasive way possible, we will refactor the upstream code to allow the carry-patch to be minimal and self-contained. An existing PoC [is available](https://github.com/openshift/kubernetes/pull/601) to demonstrate how this would look like.
+
+This patch would be carried over in OCP until all in-tree storage plugins are migrated to CSI, which should happen around 3 or 4 releases.
+Initially we would start with Cinder and GCP (_CSIMigrationGCE_ and _CSIMigrationCinder_ flags), so that we are aligned with the goals of both upstream and [CCMO](https://github.com/openshift/enhancements/pull/463).
+In subsequent OCP releases we would proceed with EBS, vSphere and Azure (_CSIMigrationAWS_, _CSIMigrationvSphere_ and _CSIMigrationAzureFile_), without a predefined order.
+
+With this patch, when deciding about using either the CSI driver or the in-tree plugin, the Attach Detach Controller will **only** rely on the information propagated by the node. In other words, Attach Detach Controller will start considering which plugin to use (in-tree or CSI) on a node basis only.
+
+Note: one might wonder why upstream does not implement this behaviour already. The answer is that they do not have the same issues OCP has in regards to disabling feature flags in the wrong order across components.
```

Comment:
why don't they have these problems? I would argue they have the same issues, but maybe ignore them. I mean, who is upstream after all? Not only GKE, AKS and EKS.

Has this problem and solution been proposed upstream?

Reply:
OCP does not follow [Kubernetes version skew policy](https://kubernetes.io/docs/setup/release/version-skew-policy/) on downgrade - OCP (CVO) downgrades API server first, *then* KCM and *then* nodes. Therefore, older KCM runs with newer kubelets for some time (until they're drained & restarted). *This is forbidden.*

Other Kubernetes vendors follow the policy, therefore they're fine. There is no issue upstream, only in OCP.

Reply:
There is a note about this few paragraphs below ("this approach should not break downgrades once CSI Migration becomes GA...")

Reply:
Spoke with @jsafrane and @bertinatto this approach is signed off by the workloads team.